templates/template.py

```diff
@@ -0,0 +1,148 @@
+    nonprint = -1      : The character is not printable.
+    combining = -2     : The character is a zero-width combiner.
+    ambiguous = -3     : The character is East-Asian ambiguous width.
+    private_use = -4   : The character is for private use.
+    unassigned = -5    : The character is unassigned.
+    widened_in_9 = -6  : Width is 1 in Unicode 8, 2 in Unicode 9+.
+    non_character = -7 : The character is a noncharacter.
+    """
+
+    nonprint = -1  # The character is not printable.
+    combining = -2  # The character is a zero-width combiner.
+    ambiguous = -3  # The character is East-Asian ambiguous width.
+    private_use = -4  # The character is for private use.
+    unassigned = -5  # The character is unassigned.
+    widened_in_9 = -6  # Width is 1 in Unicode 8, 2 in Unicode 9+.
+    non_character = -7  # The character is a noncharacter.
+
+
+class Codepointrange:
+    """A list of ranges of codepoints."""
+
+    def __init__(self, *ranges):
+        self.ranges = sorted(ranges)
+
+    def __contains__(self, char):
+        left = 0
+        right = len(self.ranges) - 1
+        if char < self.ranges[0][0] or char > self.ranges[right][1]:
+            return False
+
+        while right >= left:
+            middle = (left + right) // 2
+
+            if char < self.ranges[middle][0]:
+                right = middle - 1
+            elif char > self.ranges[middle][1]:
+                left = middle + 1
+            else:
+                return True
+        return False
+
+
+_TABLE = {{
+    # Simple ASCII characters - used a lot, so we check them first.
+    "ascii": Codepointrange(
+        {ascii}
+    ),
+    # Private usage range.
+    "private": Codepointrange(
+        {private}
+    ),
+    # Nonprinting characters.
+    "nonprint": Codepointrange(
+        {nonprint}
+    ),
+    # Width 0 combining marks.
+    "combining": Codepointrange(
+        {combining}
+    ),
+    # Width 0 combining letters.
+    "combiningletters": Codepointrange(
+        {combiningletters}
+    ),
+    # Width 2 characters.
+    "doublewide": Codepointrange(
+        {doublewide}
+    ),
+    # Ambiguous-width characters.
+    "ambiguous": Codepointrange(
+        {ambiguous}
+    ),
+    # Unassigned characters.
+    "unassigned": Codepointrange(
+        {unassigned}
+    ),
+    # Non-characters.
+    "nonchar": Codepointrange(
+        {noncharacters}
+    ),
+    # Characters that were widened from width 1 to 2 in Unicode 9.
+    "widened": Codepointrange(
+        {widenedin9}
+    ),
+}}
+
+
+# Return the width of character c, or a special negative value.
+def wcwidth(c: Union[str, int]) -> Union[int, Special]:
+    """Given a Unicode codepoint as a string or int,
+    returns the number of cells it should occupy in fixed-cell rendering like in a terminal,
+    or a negative special value (from the `Special` enum) to be handled outside.
+    """
+    if isinstance(c, str):
+        try:
+            c = ord(c)
+        except:
```

Comment:
Should be `except TypeError:`, otherwise [this will trigger for other exceptions](https://www.flake8rules.com/rules/E722.html) including KeyboardInterrupt.

Reply:
I made it `except Exception:` because I'm not sure *what* ord() actually throws - it doesn't seem to be documented. (KeyboardInterrupt and such inherit from Exception's parent, not Exception to facilitate this usage)